apps/vmq_server/src/vmq_mqtt5_fsm.erl

```diff
@@ -1376,11 +1376,19 @@ handle_messages([{deliver, QoS, Msg} = Obj|Rest], Frames, PubCnt, State, Waiting
             {Frame, NewState} = prepare_frame(QoS, Msg, State#state{fc_send_cnt=Cnt}),
             handle_messages(Rest, [Frame|Frames], PubCnt + 1, NewState, Waiting)
     end;
-handle_messages([{deliver_pubrel, {MsgId, #mqtt5_pubrel{} = Frame}}|Rest], Frames, PubCnt, State0, Waiting) ->
+handle_messages([{deliver_pubrel, {MsgId, #mqtt5_pubrel{} = Frame}}|Rest], Frames, PubCnt,
+                #state{next_msg_id = CurrNextMsgId} = State0, Waiting) ->
     %% this is called when a pubrel is retried after a client reconnects
     #state{waiting_acks=WAcks} = State0,
     _ = vmq_metrics:incr(?MQTT5_PUBREL_SENT),
-    State1 = State0#state{waiting_acks=maps:put(MsgId, Frame, WAcks)},
+    %% Assumption: all `deliver_pubrel` messages are delivered from
+    %% the queue when a session is added *before* any other messages.
+    %% The pubrel messages has fixed message-ids and we therefore need
+    %% ensure the first message id assigned to a publish message is
+    %% larger than any of the pubrel messages to avoid conflicts.
+    NextMsgId = lists:max([MsgId + 1, CurrNextMsgId]),
```

Comment:
The `MsgId + 1` could theoretically overrun the message id limit `max(ushort)`.

Reply:
Good point.. Will fix that.